Chem dispenser upgrade boards

vuonojenmustaturska opened the pull request:
Adds two new items, "chem dispenser add-in board" and "chem dispenser dangerous recipes authorization board."

The first one requires materials, data and engineering 4 to fab from the circuit imprinter, costing 1000 glass, 500 diamond and 500 gold. It effectively replicates the femto-manipulator upgrade of a constructable chem dispenser, allowing chemistry to receive a quality-of-life upgrade to their machinery that doesn't involve spacing the machines they start with and replacing them with portable ones.

The second one requires illegal tech 1 along with materials, data and engineering 4. It's effectively a single-use, chem-dispenser-only emag that requires hos or cap to bless its use. I've on purpose exluded the RD from this list. 

There are balance considerations for either, and the numbers may have to be tweaked even if the feature is approved.

[Changelogs]: 
[]: 


:cl: Naksu
add: Adds two new consumable upgrade boards for the (non-portable) chem dispenser. The first one adds the recipes a portable dispenser gains from the femto-manipulator upgrade, the other one requires authorization by the HoS or Captain and essentially emags the dispenser. These boards can be made from the circuit imprinter.
/:cl:

[why]: 
Auth board: emagged reagents have civilian uses, particularly toxin, and are not otherwise obtainable in
useful quantities.
Supplementary recipes board: the chem dispensers chemistry spawns with are non-upgradeable "due to reasons", yet allowing them to dispense "tier 4" reagents is a relatively big quality-of-life upgrade for the chemist.

vuonojenmustaturska committed:
Preliminary commit for code review

Adds two new consumable upgrade items for non-portable chem dispensers:
* Supplementary recipes add-in board adds in all the recipes a portable chem dispenser gains from a femto manipulator
* Dangerous recipes authorization boards essentially emags a non-portable chem disponsers, its use has to be blessed with an ID having either captain or HoS access

vuonojenmustaturska committed:
Changes based on code review

Previous code reviewed by Cyberboss

vuonojenmustaturska committed:
tiny fix

one more minifix

Cyberboss commented:
Ignore my out of date reviews

Cyberboss commented on code/modules/reagents/chemistry/machinery/chem_dispenser_upgrades.dm:
```diff
@@ -0,0 +1,30 @@
+/obj/item/device/chem_dispenser_auth_board/attackby(obj/item/I, mob/user, params)
+	if (istype(I, /obj/item/card/id) && check_access(I))
+		to_chat(user, "You authorize the use of [src].")
+		authorized = TRUE
```
Should probably be a different message if it was already authed

Cyberboss commented on code/modules/reagents/chemistry/machinery/chem_dispenser.dm:
```diff
@@ -195,6 +204,31 @@
+			upgraded = TRUE
+			return
+		else
+			return ..()
```
Should probably show a different message if it was already upgraded

vuonojenmustaturska committed:
Requested changes

Adds messages to user when the attempted operation (board authorization/insertion) has already been completed.

Repukan commented:
Second one seems horribly unbalanced.

Cyberboss commented:
Yeah, it seems more like a job for an RnD upgrade rather that a straight board.

Repukan commented:
I meant more in the sense that this opens a Pandora box in non traitor rounds if it only requires illegal research 1

vuonojenmustaturska commented:
The second one does require either the HoS or the Captain to "bless" the emagging, and the very reason why I made it was to have access to emag reagents in non-traitor rounds. Toxin in particular is a component of plant-b-gone, a solution to kudzu. 

I agree that the best way to accomplish the first feature would be to make the chemistry chem dispensers R&D upgradeable, but I've been told they're intentionally not because "reasons." Such work should probably include a refactor where non-portable chem dispensers are deleted and the balance issue around freshly constructed chem dispensers' reagent selection is solved somehow, either by adding a pico-manipulator to mapstart chem dispensers or putting in a non-constructable, non-researchable add-in board to function very much the same.

Repukan commented:
There are ways of getting toxin in game.

The requirement for what is opening a lot of more deadly chem mixes and one stun via miners salve doesn't quite make up being able to make plant b gone when crates can be ordered to counteract kudzu and other prs were merged to tone down it's power.

vuonojenmustaturska commented:
The numbers can and should be tweaked, I am not in any way attached to them. If only one of the ideas is implementable the other can be carved from this PR.

Anonmare commented:
Illegal research 1 begins at 1 by default roundstart FYI
